{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "reflected-canberra",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "from bs4 import BeautifulSoup\n",
    "import pandas as pd\n",
    "import time\n",
    "import shutil\n",
    "import requests\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faced-unemployment",
   "metadata": {},
   "source": [
    "# get figures from pm id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "hollywood-batch",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_article_url = 'https://www.ncbi.nlm.nih.gov/pmc/articles/'\n",
    "not_found_ids = []\n",
    "no_pathway_papers = []\n",
    "\n",
    "def create_no_pathway_file():\n",
    "    \n",
    "    with open('no_pathway.txt', 'w') as f:\n",
    "        for id in no_pathway_papers:\n",
    "            f.write(\"%s\\n\" % id)\n",
    "\n",
    "def create_not_found_file():\n",
    "    \n",
    "    with open('not_found_ids.txt', 'w') as f:\n",
    "        for id in not_found_ids:\n",
    "            f.write(\"%s\\n\" % id)\n",
    "\n",
    "def pubmed_id_to_PMC(id):\n",
    "    \n",
    "    a = requests.get(\"https://www.ncbi.nlm.nih.gov/pmc/utils/idconv/v1.0/?ids=\" + str(id) + \"&format=json\")\n",
    "    all_id =  a.json()\n",
    "    all_list = all_id.get('records')\n",
    "    b = list(all_list[0].values())\n",
    "    \n",
    "    if b[0][:3] != 'PMC':\n",
    "        not_found_ids.append(id)\n",
    "    \n",
    "    return b[0]\n",
    "\n",
    "def get_page_soup(PMC):\n",
    "    \n",
    "    page = requests.get(base_article_url + PMC + '/')\n",
    "    soup = BeautifulSoup(page.content, 'html.parser')\n",
    "    \n",
    "    return soup\n",
    "\n",
    "def get_captions(soup):\n",
    "    \n",
    "    caption_elements = soup.find_all(lambda tag: tag.name == 'div' and tag.get('class') == ['caption'])\n",
    "    captions = [caption_element.findChild().text for caption_element in  caption_elements]\n",
    "    \n",
    "    return captions\n",
    "\n",
    "def get_figure_numbers(soup):\n",
    "    \n",
    "    img_elements = soup.find_all(lambda tag: tag.name == 'a' and tag.get('class') == ['figpopup'])\n",
    "    fig_numbers = [img_element.text for img_element in img_elements]\n",
    "    \n",
    "    return fig_numbers\n",
    "    \n",
    "def get_img_names(soup):\n",
    "       \n",
    "    img_elements = soup.find_all('img', class_ = 'tileshop')\n",
    "    sources = [img_element['src'] for img_element in img_elements]\n",
    "    after_bins = [re.search(\"(bin(.*))\", source).group() for source in sources]\n",
    "    \n",
    "    return after_bins\n",
    "\n",
    "def get_image(PMC, source, figure_number, id):\n",
    "\n",
    "    response = requests.get(base_article_url + PMC + '/' + source, stream=True)\n",
    "    \n",
    "    save_path = 'output/'+ str(id)\n",
    "    try:\n",
    "        os.makedirs(save_path)\n",
    "    except:\n",
    "        print(save_path + ' exists')\n",
    "    \n",
    "    with open(save_path + '/' + figure_number + '.jpg', 'wb') as out_file:\n",
    "        shutil.copyfileobj(response.raw, out_file)\n",
    "        \n",
    "def take_imgs_from_site(id):\n",
    "    \n",
    "    PMC = pubmed_id_to_PMC(id)\n",
    "    soup = get_page_soup(PMC)   \n",
    "    \n",
    "    captions = get_captions(soup)\n",
    "    sources = get_img_names(soup)\n",
    "    figure_numbers = get_figure_numbers(soup)\n",
    "    \n",
    "    pathway_found_counter = 0\n",
    "    for source, figure_number, caption in zip(sources, figure_numbers, captions):\n",
    "        if ('athway' or 'etwork') in caption:\n",
    "            get_image(PMC, source, figure_number, id)\n",
    "            pathway_found_counter+=1\n",
    "    \n",
    "    if pathway_found_counter == 0:\n",
    "        no_pathway_papers.append(id)\n",
    "    \n",
    "    print('Paper id ' + str(id) + ' done')\n",
    "\n",
    "def take_imgs_from_multiple_sites(ids):\n",
    "    \n",
    "    id_counter = 1\n",
    "    for id in ids:\n",
    "        \n",
    "        if id.isnumeric():\n",
    "            take_imgs_from_site(id)\n",
    "        print(str(len(ids)-id_counter) + ' ids left')\n",
    "        id_counter+=1\n",
    "    \n",
    "    create_no_pathway_file()\n",
    "    create_not_found_file()\n",
    "    print('All papers done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "sustained-maker",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paper id 32145363 done\n",
      "142 ids left\n",
      "Paper id 31226023 done\n",
      "141 ids left\n",
      "Paper id 32330414 done\n",
      "140 ids left\n",
      "Paper id 32511329 done\n",
      "139 ids left\n",
      "Paper id 15194747 done\n",
      "138 ids left\n",
      "Paper id 32132184 done\n",
      "137 ids left\n",
      "Paper id 31034780 done\n",
      "136 ids left\n",
      "Paper id 15331731 done\n",
      "135 ids left\n",
      "Paper id 16877062 done\n",
      "134 ids left\n",
      "Paper id 32225175 done\n",
      "133 ids left\n",
      "Paper id 32227760 done\n",
      "132 ids left\n",
      "131 ids left\n",
      "Paper id 33899609 done\n",
      "130 ids left\n",
      "Paper id 33482803 done\n",
      "129 ids left\n",
      "Paper id 32265513 done\n",
      "128 ids left\n",
      "Paper id 32219363 done\n",
      "127 ids left\n",
      "Paper id 32976572 done\n",
      "126 ids left\n",
      "Paper id 33436497 done\n",
      "125 ids left\n",
      "Paper id 32353859 done\n",
      "124 ids left\n",
      "Paper id 32358202 done\n",
      "123 ids left\n",
      "Paper id 32346093 done\n",
      "122 ids left\n",
      "Paper id 32835247 done\n",
      "121 ids left\n",
      "Paper id 32227090 done\n",
      "120 ids left\n",
      "Paper id 32398875 done\n",
      "119 ids left\n",
      "output/32522207 exists\n",
      "Paper id 32522207 done\n",
      "118 ids left\n",
      "Paper id 32835326 done\n",
      "117 ids left\n",
      "Paper id 32546811 done\n",
      "116 ids left\n",
      "output/28933406 exists\n",
      "output/28933406 exists\n",
      "Paper id 28933406 done\n",
      "115 ids left\n",
      "Paper id 12441076 done\n",
      "114 ids left\n",
      "Paper id 22013052 done\n",
      "113 ids left\n",
      "Paper id 19109397 done\n",
      "112 ids left\n",
      "Paper id 26203058 done\n",
      "111 ids left\n",
      "Paper id 22553325 done\n",
      "110 ids left\n",
      "109 ids left\n",
      "Paper id 32587962 done\n",
      "108 ids left\n",
      "Paper id 33206959 done\n",
      "107 ids left\n",
      "Paper id 18753196 done\n",
      "106 ids left\n",
      "Paper id 10799570 done\n",
      "105 ids left\n",
      "Paper id 22751485 done\n",
      "104 ids left\n",
      "Paper id 15527857 done\n",
      "103 ids left\n",
      "Paper id 32048163 done\n",
      "102 ids left\n",
      "Paper id 32591762 done\n",
      "101 ids left\n",
      "Paper id 32663912 done\n",
      "100 ids left\n",
      "Paper id 32732413 done\n",
      "99 ids left\n",
      "Paper id 32778050 done\n",
      "98 ids left\n",
      "Paper id 32807934 done\n",
      "97 ids left\n",
      "Paper id 32946807 done\n",
      "96 ids left\n",
      "Paper id 32714119 done\n",
      "95 ids left\n",
      "Paper id 32511375 done\n",
      "94 ids left\n",
      "Paper id 32270184 done\n",
      "93 ids left\n",
      "Paper id 32386571 done\n",
      "92 ids left\n",
      "Paper id 32763915 done\n",
      "91 ids left\n",
      "90 ids left\n",
      "89 ids left\n",
      "output/32633718 exists\n",
      "Paper id 32633718 done\n",
      "88 ids left\n",
      "Paper id 32968798 done\n",
      "87 ids left\n",
      "86 ids left\n",
      "85 ids left\n",
      "84 ids left\n",
      "Paper id 33046644 done\n",
      "83 ids left\n",
      "Paper id 32877502 done\n",
      "82 ids left\n",
      "Paper id 33082294 done\n",
      "81 ids left\n",
      "Paper id 33082293 done\n",
      "80 ids left\n",
      "Paper id 32528623 done\n",
      "79 ids left\n",
      "Paper id 32869019 done\n",
      "78 ids left\n",
      "output/33157038 exists\n",
      "Paper id 33157038 done\n",
      "77 ids left\n",
      "Paper id 33116139 done\n",
      "76 ids left\n",
      "Paper id 33147445 done\n",
      "75 ids left\n",
      "Paper id 29642015 done\n",
      "74 ids left\n",
      "Paper id 33125439 done\n",
      "73 ids left\n",
      "Paper id 33116300 done\n",
      "72 ids left\n",
      "Paper id 33147444 done\n",
      "71 ids left\n",
      "Paper id 33184506 done\n",
      "70 ids left\n",
      "69 ids left\n",
      "Paper id 33278357 done\n",
      "68 ids left\n",
      "Paper id 32142651 done\n",
      "67 ids left\n",
      "Paper id 32944968 done\n",
      "66 ids left\n",
      "Paper id 33244168 done\n",
      "65 ids left\n",
      "Paper id 33270413 done\n",
      "64 ids left\n",
      "Paper id 33060197 done\n",
      "63 ids left\n",
      "Paper id 33277466 done\n",
      "62 ids left\n",
      "output/33290397 exists\n",
      "Paper id 33290397 done\n",
      "61 ids left\n",
      "Paper id 33139913 done\n",
      "60 ids left\n",
      "Paper id 32972995 done\n",
      "59 ids left\n",
      "Paper id 32972996 done\n",
      "58 ids left\n",
      "Paper id 33307546 done\n",
      "57 ids left\n",
      "Paper id 33188728 done\n",
      "56 ids left\n",
      "Paper id 33296687 done\n",
      "55 ids left\n",
      "Paper id 33159009 done\n",
      "54 ids left\n",
      "Paper id 33324374 done\n",
      "53 ids left\n",
      "52 ids left\n",
      "Paper id 33402434 done\n",
      "51 ids left\n",
      "Paper id 33422265 done\n",
      "50 ids left\n",
      "Paper id 33231615 done\n",
      "49 ids left\n",
      "output/33253796 exists\n",
      "output/33253796 exists\n",
      "output/33253796 exists\n",
      "Paper id 33253796 done\n",
      "48 ids left\n",
      "Paper id 33451855 done\n",
      "47 ids left\n",
      "Paper id 32191739 done\n",
      "46 ids left\n",
      "Paper id 32416070 done\n",
      "45 ids left\n",
      "Paper id 32645325 done\n",
      "44 ids left\n",
      "Paper id 33547084 done\n",
      "43 ids left\n",
      "Paper id 33576820 done\n",
      "42 ids left\n",
      "Paper id 33576814 done\n",
      "41 ids left\n",
      "Paper id 33495306 done\n",
      "40 ids left\n",
      "Paper id 33536217 done\n",
      "39 ids left\n",
      "Paper id 14607900 done\n",
      "38 ids left\n",
      "output/33503446 exists\n",
      "Paper id 33503446 done\n",
      "37 ids left\n",
      "Paper id 33590968 done\n",
      "36 ids left\n",
      "Paper id 33361116 done\n",
      "35 ids left\n",
      "Paper id 33732251 done\n",
      "34 ids left\n",
      "Paper id 33655830 done\n",
      "33 ids left\n",
      "Paper id 32967005 done\n",
      "32 ids left\n",
      "Paper id 33686287 done\n",
      "31 ids left\n",
      "Paper id 33524849 done\n",
      "30 ids left\n",
      "Paper id 33492978 done\n",
      "29 ids left\n",
      "Paper id 16125763 done\n",
      "28 ids left\n",
      "Paper id 29245018 done\n",
      "27 ids left\n",
      "Paper id 33713620 done\n",
      "26 ids left\n",
      "Paper id 33727703 done\n",
      "25 ids left\n",
      "Paper id 33723416 done\n",
      "24 ids left\n",
      "Paper id 33328588 done\n",
      "23 ids left\n",
      "22 ids left\n",
      "Paper id 32205204 done\n",
      "21 ids left\n",
      "Paper id 33348292 done\n",
      "20 ids left\n",
      "Paper id 33211371 done\n",
      "19 ids left\n",
      "Paper id 33811162 done\n",
      "18 ids left\n",
      "Paper id 33784499 done\n",
      "17 ids left\n",
      "Paper id 33984267 done\n",
      "16 ids left\n",
      "Paper id 34029205 done\n",
      "15 ids left\n",
      "Paper id 34010139 done\n",
      "14 ids left\n",
      "Paper id 33845483 done\n",
      "13 ids left\n",
      "Paper id 33879890 done\n",
      "12 ids left\n",
      "Paper id 33880470 done\n",
      "11 ids left\n",
      "Paper id 33915569 done\n",
      "10 ids left\n",
      "Paper id 33976134 done\n",
      "9 ids left\n",
      "Paper id 33915568 done\n",
      "8 ids left\n",
      "Paper id 32733001 done\n",
      "7 ids left\n",
      "Paper id 32979938 done\n",
      "6 ids left\n",
      "output/32645325 exists\n",
      "Paper id 32645325 done\n",
      "5 ids left\n",
      "Paper id 32529952 done\n",
      "4 ids left\n",
      "Paper id 33451855 done\n",
      "3 ids left\n",
      "Paper id 32589897 done\n",
      "2 ids left\n",
      "Paper id 32555321 done\n",
      "1 ids left\n",
      "0 ids left\n",
      "All papers done\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('Collected papers and classification proposal - Collected papers and classification proposal.csv')\n",
    "ids = df['Pubmed ID'].tolist()\n",
    "\n",
    "take_imgs_from_multiple_sites(ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "miniature-psychiatry",
   "metadata": {},
   "source": [
    "# Get figures from pmid that do not have PMCID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "broad-quarter",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_article_url = 'https://pubmed.ncbi.nlm.nih.gov/'\n",
    "not_found_ids = []\n",
    "no_pathway_papers = []\n",
    "\n",
    "def create_no_pathway_file():\n",
    "\n",
    "    with open('no_pathway_from_no_pmcids.txt', 'w') as f:\n",
    "        for id in no_pathway_papers:\n",
    "            f.write(\"%s\\n\" % id)\n",
    "\n",
    "def get_page_soup(pmid):\n",
    "    \n",
    "    page = requests.get(base_article_url + str(pmid) + '/')\n",
    "    soup = BeautifulSoup(page.content, 'html.parser')\n",
    "    \n",
    "    return soup\n",
    "\n",
    "def get_fulltext_link(soup):\n",
    "       \n",
    "    full_text_links = soup.find_all(lambda tag: tag.name == 'a' and tag.get('class') == ['id-link'])\n",
    "    links = [full_text_link['href'] for full_text_link in full_text_links]\n",
    "    \n",
    "    return links[0]\n",
    "\n",
    "def get_full_text_soup(links):\n",
    "    \n",
    "    text_page = requests.get(links)\n",
    "    text_soup = BeautifulSoup(text_page.content, 'html.parser')\n",
    "    \n",
    "    return text_soup\n",
    "\n",
    "def get_fig_captions_pmid(text_soup):\n",
    "    \n",
    "    figcaption_codes = text_soup.find_all(lambda tag: tag.name == 'b' and tag.get('class') == ['c-article-section__figure-caption'])\n",
    "    #captions = [caption_element.text for caption_element in fig]\n",
    "    #figcaption_codes = soup.find_all('figcaption')\n",
    "    \n",
    "    return figcaption_codes\n",
    "            \n",
    "###############\n",
    "def get_img_sources(soup):\n",
    "       \n",
    "    #img_containers = soup.find_all('a', class_ = 'ar-figure-viewer-modal figure scrollFig')\n",
    "    #img_srcs = [img_container.findChild()['src'] for img_container in  img_containers]\n",
    "      \n",
    "    #img_containers = soup.find_all('a', class_ = 'highwire-figure-link highwire-figure-link-download')  \n",
    "    #img_srcs = [img_container['href'] for img_container in img_containers]\n",
    "    \n",
    "    img_containers = soup.find_all('a', class_ = 'c-article-section__figure-link')\n",
    "    img_srcs = [img_container['href'] for img_container in img_containers]\n",
    "    return img_srcs\n",
    "\n",
    "def get_image(source,fake_fig_num, id):\n",
    "    \n",
    "    response = requests.get('https://www.annualreviews.org/'+ source, stream=True)\n",
    "    #response = requests.get(source)\n",
    "    save_path = 'output_noPMCIDs/'+ str(id)\n",
    "    try:\n",
    "        os.makedirs(save_path)\n",
    "    except:\n",
    "        print(save_path + ' exists')\n",
    "    \n",
    "    #with open(save_path + '/' + caption + '.jpg', 'wb') as out_file:\n",
    "       # shutil.copyfileobj(response.raw, out_file)         \n",
    "    \n",
    "  \n",
    "    file = open(save_path + '/' + str(fake_fig_num) + source[-4:], \"wb\")\n",
    "    file.write(response.content)\n",
    "    file.close()\n",
    "###############\n",
    "\n",
    "def get_from_one_id(id):\n",
    "    soup = get_page_soup(id)\n",
    "    links = get_fulltext_link(soup)\n",
    "    text_soup = get_full_text_soup(links)\n",
    "    caption_codes = get_fig_captions_pmid(text_soup)\n",
    "    img_sources = get_img_sources(text_soup)\n",
    "    \n",
    "    print(caption_codes)\n",
    "    \n",
    "    pathway_found_counter = 0\n",
    "    for source, caption_code, fake_fig_num in zip(img_sources, caption_codes, range(1, len(img_sources) + 1)):\n",
    "        if ('athway' or 'etwork') in caption_code:\n",
    "            get_image(source, caption_code, fake_fig_num, id)\n",
    "            pathway_found_counter += 1\n",
    "     \n",
    "    #for source, fake_fig_num in zip(img_sources, range(1, len(img_sources) + 1)):\n",
    "        #get_image(source, fake_fig_num, id)\n",
    "        #pathway_found_counter += 1\n",
    "            \n",
    "    if pathway_found_counter == 0:\n",
    "        no_pathway_papers.append(id)\n",
    "    \n",
    "    print('Paper id ' + str(id) + ' done')    \n",
    "    \n",
    "def get_from_multi_ids(ids):\n",
    "    id_counter = 1\n",
    "    for id in ids:\n",
    "        get_from_one_id(id)\n",
    "        print(str(len(ids)-id_counter) + ' ids left')\n",
    "        id_counter+=1\n",
    "        \n",
    "    create_no_pathway_file()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "understood-japan",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "Paper id 31226023 done\n",
      "24 ids left\n",
      "[]\n",
      "Paper id 32219363 done\n",
      "23 ids left\n",
      "[<b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig1\">Fig. 1: Dysregulated bronchoalveolar immune landscapes in patients with severe COVID-19 infection.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig2\">Fig. 2: T cell clonal expansion in BALF from patients with COVID-19.</b>]\n",
      "Paper id 32398875 done\n",
      "22 ids left\n",
      "[]\n",
      "Paper id 32546811 done\n",
      "21 ids left\n",
      "[<b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig1\">Fig. 1: Illustration of the experimental setup and the patient cohort used in this study.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig2\">Fig. 2: Identification, distribution and classification of cells from the upper respiratory tract.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig3\">Fig. 3: Inferred differentiation pathways for epithelial cells from the upper respiratory tract.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig4\">Fig. 4: Immune–epithelial cell interaction in COVID-19.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig5\">Fig. 5: Differences between moderate and critical COVID-19 severity are attributed to non-resident macrophage over-expression of cytokines/chemokines and extravasation.</b>]\n",
      "Paper id 32591762 done\n",
      "20 ids left\n",
      "[]\n",
      "Paper id 32732413 done\n",
      "19 ids left\n",
      "[<b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig1\">Fig. 1: Immunophenotyping reveals distinct features of the immune system in patients with COVID-19.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig2\">Fig. 2: Coexistent anti-SARS-CoV-2 antibody responses and disruption to the B cell compartment in patients with COVID-19.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig3\">Fig. 3: Patients with COVID-19 display dysregulated cytokine responses.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig4\">Fig. 4: Patients with COVID-19 display a disrupted monocyte and dendritic cell phenotype.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig5\">Fig. 5: Patients with COVID-19 display selective cytopenia in particular T cell subsets and coexistent active cycling and exhausted phenotype of residual T cells.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig6\">Fig. 6: COVID-IP parameters associated with patient prognosis.</b>]\n",
      "Paper id 32807934 done\n",
      "18 ids left\n",
      "[]\n",
      "Paper id 33082294 done\n",
      "17 ids left\n",
      "[<b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig1\">Fig. 1: The S protein of SARS-CoV-2 binds to cholesterol.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig2\">Fig. 2: HDL enhances the entry of SARS-CoV-2 through SARS-2-S1 protein.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig3\">Fig. 3: SR-B1 expression confers susceptibility to SARS-CoV-2 infection.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig4\">Fig. 4: SR-B1 mediates SARS-CoV-2 attachment and entry.</b>]\n",
      "Paper id 33244168 done\n",
      "16 ids left\n",
      "[]\n",
      "Paper id 33270413 done\n",
      "15 ids left\n",
      "[]\n",
      "Paper id 33139913 done\n",
      "14 ids left\n",
      "[<b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig1\">Fig. 1: Discovery GWAS and meta-analysis results.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig2\">Fig. 2: Summary of TWAS results.</b>]\n",
      "Paper id 33307546 done\n",
      "13 ids left\n",
      "[]\n",
      "Paper id 33576814 done\n",
      "12 ids left\n",
      "[]\n",
      "Paper id 33536217 done\n",
      "11 ids left\n",
      "[]\n",
      "Paper id 14607900 done\n",
      "10 ids left\n",
      "[]\n",
      "Paper id 33655830 done\n",
      "9 ids left\n",
      "[<b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig1\">Fig. 1: Genome-wide knockout screens in human cells identify host factors for SARS-CoV-2 and HCoV-229E infection.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig2\">Fig. 2: Validation of genetic screens.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig3\">Fig. 3: PI3K type 3 is a druggable target against SARS-CoV-2 and other coronaviruses.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig4\">Fig. 4: TMEM41B is a host factor for HCoV-229E.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig5\">Fig. 5: <i>TMEM106B</i> is a bona fide proviral host factor for SARS-CoV-2 replication.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig6\">Fig. 6: <i>TMEM106B</i> expression in epithelial cells from human airways correlates with SARS-CoV-2 infection.</b>]\n",
      "Paper id 33686287 done\n",
      "8 ids left\n",
      "[]\n",
      "Paper id 33492978 done\n",
      "7 ids left\n",
      "[]\n",
      "Paper id 16125763 done\n",
      "6 ids left\n",
      "[]\n",
      "Paper id 29245018 done\n",
      "5 ids left\n",
      "[<b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig1\">Fig. 1: Clofazimine inhibits the replication of a broad spectrum of human-pathogenic coronaviruses in human cellular models.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig2\">Fig. 2: Clofazimine interferes with several steps of the virus life cycle.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig3\">Fig. 3: Prophylactic and therapeutic treatment with clofazimine reduces disease in a hamster model of SARS-CoV-2 infection.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig4\">Fig. 4: Clofazimine exhibits antiviral synergy with remdesivir in hamsters.</b>]\n",
      "Paper id 33727703 done\n",
      "4 ids left\n",
      "[]\n",
      "Paper id 33811162 done\n",
      "3 ids left\n",
      "[<b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig1\">Fig. 1: Joint analysis of SARS-CoV-2 and SARS-CoV virus–host protein–protein interactomes.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig2\">Fig. 2: Multi-level profiling of SARS-CoV-2 and SARS-CoV infection.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig3\">Fig. 3: Phosphorylation and ubiquitination of SARS-CoV-2 and SARS-CoV viral proteins.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig4\">Fig. 4: A network diffusion approach enables identification of molecular pathways linking protein–protein interactions with downstream changes in the host proteome.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig5\">Fig. 5: Identification of pathways targeted by SARS-CoV-2 using a multi-omics profiling approach enables systematic testing of candidate antiviral therapies.</b>]\n",
      "Paper id 33845483 done\n",
      "2 ids left\n",
      "[<b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig1\">Fig. 1: Experimental and computational pipeline for a COVID-19 autopsy atlas.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig2\">Fig. 2: A single-cell and single-nucleus atlas of COVID-19 lung.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig3\">Fig. 3: SARS-CoV-2 RNA<sup>+</sup> single cells are enriched for phagocytic and endothelial cells.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig4\">Fig. 4: Composition and expression differences between COVID-19 and healthy lungs and between infected and uninfected regions within COVID-19 lungs.</b>]\n",
      "Paper id 33915569 done\n",
      "1 ids left\n",
      "[<b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig1\">Fig. 1: Study design and cellular landscape.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig2\">Fig. 2: Immune responses in COVID-19.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig3\">Fig. 3: Impaired lung regeneration and sources of inflammation.</b>, <b class=\"c-article-section__figure-caption\" data-test=\"figure-caption-text\" id=\"Fig4\">Fig. 4: Pathological fibroblasts and ensuing fibrosis in COVID-19.</b>]\n",
      "Paper id 33915568 done\n",
      "0 ids left\n"
     ]
    }
   ],
   "source": [
    "ids = open('not_found_ids.txt').read().splitlines()\n",
    "get_from_multi_ids(ids)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
